app/drivers/adc.c

```diff
@@ -43,9 +57,66 @@ err_t adc_init(void)
 	status = HAL_ADC_ConfigChannel(&adc_handle, &adc_config);
 	HAL_ERR_CHECK(status, EADC_HAL_CONFIG_CHANNEL);
 
+	adc_config.Channel = ADC_CHANNEL_6;
+	adc_config.Rank = ADC_REGULAR_RANK_2;
+	status = HAL_ADC_ConfigChannel(&adc_handle, &adc_config);
+	HAL_ERR_CHECK(status, EADC_HAL_CONFIG_CHANNEL);
+
+	m_initialized = true;
+
 	return ERR_OK;
 }
 
+err_t adc_start(void)
+{
+	if (!m_initialized)
+		return EADC_NO_INIT;
+
+	HAL_NVIC_EnableIRQ(ADC1_2_IRQn);
+	HAL_ADC_Start_IT(&adc_handle);
+
+	return ERR_OK;
+}
+
+err_t adc_stop(void)
+{
+	if (!m_initialized)
+		return EADC_NO_INIT;
+
+	HAL_ADC_Stop_IT(&adc_handle);
+	HAL_NVIC_DisableIRQ(ADC1_2_IRQn);
+
+	return ERR_OK;
+}
+
+/* Note that callback is called from interrupt context */
+void adc_set_callback(adc_conversion_ready callback)
+{
+	m_callback = callback;
+}
+
+void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef* hadc)
+{
+	m_adc_samples_count++;
+
+	/* If End Of Sequence is set, the last sample in our sequence is done.
+	 * TODO: Dig deeper in the HAL and see if register access is really necessary.
+	 */
+	if (hadc->Instance->ISR & ADC_ISR_EOS) {
+		m_adc_values[1] = HAL_ADC_GetValue(&adc_handle);
+		if (m_callback)
+			m_callback(m_adc_values);
+	}
```

Comment:
Woops

Reply:
What is wrong here? I guess it looks a bit funky but the interrupt is fired for each sampling that gets completed, which happens twice because we sample 2 channels. I check the register `hadc->Instance->ISR & ADC_ISR_EOS` in order to know if the *sequence* is complete, which in our case is when the second channel has been sampled. I guess there is a better way to do this, but i'm not sure really. Guess I could just have a counter, but I'm worried what might happen if i miss an interrupt, would not want to mix up ch5 and ch6.